src/group_theory/perm/fin.lean

```diff
@@ -62,3 +62,130 @@ lemma finset.univ_perm_fin_succ {n : ℕ} :
   @finset.univ (perm $ fin n.succ) _ = (finset.univ : finset $ fin n.succ × perm (fin n)).map
   equiv.perm.decompose_fin.symm.to_embedding :=
 (finset.univ_map_equiv_to_embedding _).symm
+
+section cycle_range
+
+/-! ### `cycle_range` section
+
+Define the permutation `fin.cycle_range i`, the cycle `(0 1 2 ... i)`.
+The definition is slightly awkward because we need to keep track of the fact
+that the cycle indeed stays within `fin n`. You're advised to use
+`cycle_range_of_lt`, `cycle_range_self` and `cycle_range_of_gt` instead of
+unfolding the definition.
+-/
+
+namespace fin
+
+/-- `fin.cycle_range i` is the cycle `(0 1 2 ... i)`. -/
+def cycle_range {n : ℕ} : fin n → perm (fin n)
+-- We could write this as a product over `list.range`,
+-- but the explicit `mk`s would become `coe`s and make the proofs a lot more annoying.
+-- We could also write this as two `if`s, but it is very annoying to prove that
+-- indeed results in a permutation.
```

Comment:
Does it make sense to break this definition into two pieces; one that does the rotation, and one that does the pass-through?
```
/-- A permutation of `fin m` that rotates elements by `n` -/
def cycle (n m : ℕ) : perm (fin m) :=
{ to_fun := λ i, i + n,  -- this would work for zmod...
  inv_fun := λ i, i - n,
  left_inv := λ i, sorry,
  right_inv := λ i, sorry, }

/-- rotate the first `m` elements by `1`, leave the remaining `n` unchanged. -/
def cycle_range (m : ℕ) (n : ℕ) : perm (fin $ m + n) :=
sum_fin_sum_equiv.perm_congr ((cycle 1 m).sum_congr (equiv.refl $ fin n))
```

Reply:
Hmm, I'm not really convinced that those are the natural ways to carve up this definition. In fact, wouldn't it be nicer to define `cycle n m` as `(cycle_range (fin.last n)) ^ m`? (Also, `cycle` is probably not the right name for that map since [`is_cycle`](https://leanprover-community.github.io/mathlib_docs/group_theory/perm/cycles.html#equiv.perm.is_cycle) has a specific meaning in group theory. If you want it, it should be something like `add_perm`.)

If we were to generalize, group theorists would prefer `cycle (xs : list α) : perm α`, sending `xs.nth 0` to `xs.nth 1`, `xs.nth 1` to `xs.nt 2`, ... `xs.nth (length xs - 1)` to `xs.nth 0`. (And `fin.cycle_range i` would just be `perm.cycle (list.range i)`.) But list indices are not so nice to work with, so you'd want to use `fin` indices instead, and the result is `cycle_range` composed with some injective `fin n -> α`.

Reply:
I guess what I'm finding weird here is the "cycle the first m, leave the rest" behavior. We have lots of combinators for "permute just these elements", what we're missing is the "cycle all elements" operation.

> But list indices are not so nice to work with, so you'd want to use fin indices instead, and the result is cycle_range composed with some injective fin n -> α.

Here, the "permute just these elements" combinator would be:
```
def subtype_congr_fin {α : Type*} {n : ℕ} (f : fin n ↪ α) [decidable_eq α] (e : perm $ fin n) : perm α :=
let e' : perm (set.range f) := equiv.perm_congr {
  to_fun := λ x, (⟨f x, x, rfl⟩ : set.range f),
  inv_fun := f.inv_of_mem_range,
  left_inv := f.right_inv_of_inv_of_mem_range,
  right_inv := λ x, subtype.eq $ f.left_inv_of_inv_of_mem_range x,
} e in
equiv.perm.subtype_congr e' (equiv.refl _)
```

Reply:
Aha, so your goal looks more like using an embedding to turn  "permute the whole of `fin n`" into "permute the `fin n` subset of `fin k`"?

Reply:
Yes, with the bonus that now we're not really tied to `fin` at all - the only thing that actually requires `fin` is the cycle sending `i → i + 1`, `fin.last → 0`

Reply:
[I managed to define](https://github.com/leanprover-community/mathlib/commit/0107630b0d4b92f424ad25f8d381a5f614b15b12#diff-61fef3e3002e0edd652a7dae93e737aa1111caf3ebb9a05cb007417f73837ceaR79) `fin.cycle_range` in terms of a new `fin.cycle_all`, but I'm really not convinced this is the way to go. (I would much rather define `fin.cycle_all` in terms of `fin.cycle_range`.) Trying to compute with these `subtype_congr` and `perm_congr` means 10s of lines of pushing around trivial inequalities, and it's really hard to get the sign of these permutations without basically resorting to the original definition of `cycle_range` anyway. Or is there some golfing trick that I'm missing?

Reply:
[I just pushed](https://github.com/leanprover-community/mathlib/pull/6815/commits/aaecafe875f61f848a2578aaa7d6c5571a0214b0) a definition of `cycle_all` and the same results, defined in terms of `cycle_range`. I definitely know which commit I prefer :)

Reply:
Here's a much shorter defnition of `cycle_all`:
```
def cycle_all : Π (n : ℕ), equiv.perm (fin n)
| 0 := 1
| (nat.succ n) := show equiv.perm (zmod n.succ), from
{ to_fun := λ i, i + 1,
  inv_fun := λ i, i - 1,
  left_inv := λ i, by simp,
  right_inv := λ i, by simp, }
```

Reply:
I wasn't able to try out your branch because of build failures in `equiv/basic`

Reply:
To make the `equiv.add_right` definition non-painful, we need #6848

Reply:
> ```lean
> -- import data.zmod.basic for fin.comm_ring
> 
> /-- `fin.cycle_all` is the cycle `(0 1 2 ... fin.last n)`. -/
> def cycle_all : Π (n : ℕ), perm (fin n)
> | 0 := 1
> | (nat.succ n) := let inst := fin.comm_ring n in by exactI equiv.add_right 1
> ```

So we standardised `fin` back on wrapping addition? That should help a bunch indeed!

Reply:
I'll see in a couple hours if I can get it to work.

Reply:
I think you can get `sign (cycle_all n)` by showing `cycle_all n.succ` is equal to `decompose_fin.symm (1, cycle_all n)` or similar.

Reply:
Took a bit longer than expected, the proofs are still not as nice as I was hoping, but I can live with this approach.